### PR TITLE
Task 1 (Abstract factory) by Tihomir Kehayov

### PR DESCRIPTION
**Pull request by tkehayov**
(abstract_factory) implementing abstract factory task using interfaces and reflection.

### CODE REVIEW

**Comment by mgenov** [comment]
Doesn't look like an abstract factory to me. Seems a little bit simpler.

**Comment by tkehayov** [comment]
Done...

File: src/com/clouway/abstractfactory/Builder.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package com.clouway.abstractfactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author Tihomir Kehayov <kehayov89@gmail.com>
+ */
+public class Builder implements HouseFactory {
+//  @Override
```
Comment (mgenov) [inline src/com/clouway/abstractfactory/Builder.java]: commented code need to be removed.

File: src/com/clouway/abstractfactory/Builder.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package com.clouway.abstractfactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author Tihomir Kehayov <kehayov89@gmail.com>
+ */
+public class Builder implements HouseFactory {
```
Comment (mgenov) [inline src/com/clouway/abstractfactory/Builder.java]: Builder is another pattern. Not an abstract factory.

**Comment by mgenov** [comment]
Still no, it doesn't look like abstract factory. An abstract factory is creating factory of factories. I don't see this in your example.

File: src/com/clouway/abstractfactory/Builder.java (hunk 1 above)
Comment (tkehayov) [inline src/com/clouway/abstractfactory/Builder.java]: done

File: src/com/clouway/abstractfactory/Builder.java (hunk 2 above)
Comment (tkehayov) [inline src/com/clouway/abstractfactory/Builder.java]: Builder is just a name

**Comment by tkehayov** [comment]
Now should be better.

P.S. Why is missing "Add a line note on last comment"?

File: src/com/clouway/abstractfactory/Builder.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package com.clouway.abstractfactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author Tihomir Kehayov <kehayov89@gmail.com>
+ */
+public class Builder {
```
Comment (mgenov) [inline src/com/clouway/abstractfactory/Builder.java]: I know that it's just a name. But when I see class named builder, I'm thinking that it's a builder, not a factory. These are 2 different things. It's same as naming dog as cat.

File: src/com/clouway/abstractfactory/Builder.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  public HouseFactory makeHouse(HouseFactory houseFactory) {
+    return houseFactory;
```
Comment (mgenov) [inline src/com/clouway/abstractfactory/Builder.java]: Builder in your case doesn't create anything, in just returns the provided house factory. The idea of abstract factory is to create factory of factories.

Comment (tkehayov) [inline src/com/clouway/abstractfactory/Builder.java]: done

File: src/com/clouway/abstractfactory/Building.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package com.clouway.abstractfactory;
+
+/**
+ * @author Tihomir Kehayov <kehayov89@gmail.com>
+ */
+public class Building implements Factory {
```
Comment (mgenov) [inline src/com/clouway/abstractfactory/Building.java]: Is more common Factory classes to end with Factory to indicate that this is factory object that creates objects.

File: src/com/clouway/abstractfactory/Demo.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+import com.clouway.abstractfactory.parts.Floor;
+import com.clouway.abstractfactory.parts.Door;
+import com.clouway.abstractfactory.parts.Window;
+
+/**
+ * @author Tihomir Kehayov <kehayov89@gmail.com>
+ */
+public class Demo {
+  public static void main(String[] args) {
+    Factory building = new Producer().create(Building.class);
```
Comment (mgenov) [inline src/com/clouway/abstractfactory/Demo.java]: Building and House are looking same, so I don't think that this is a real example of AbstractFactory. 

Note that you can move the reflection based factory method in another package to separate them. 

I've changed the task description, cause there is some kind of misunderstanding about this task.

Comment (tkehayov) [inline src/com/clouway/abstractfactory/Demo.java]: I think that now is great.

**Pull request closed without merging**